pkg/virt-handler/device-manager/usb_device.go

```diff
@@ -0,0 +1,547 @@
+			if err != nil {
+				return nil
+			}
+			u.Bus = int(val)
+		case "DEVNUM":
+			val, err := strconv.ParseInt(values[1], 10, 32)
+			if err != nil {
+				return nil
+			}
+			u.DeviceNumber = int(val)
+		case "PRODUCT":
+			products := strings.Split(values[1], "/")
+			if len(products) != 3 {
+				return nil
+			}
+
+			val, err := strconv.ParseInt(products[0], 16, 32)
+			if err != nil {
+				return nil
+			}
+			u.Vendor = int(val)
+
+			val, err = strconv.ParseInt(products[1], 16, 32)
+			if err != nil {
+				return nil
+			}
+			u.Product = int(val)
+
+			val, err = strconv.ParseInt(products[2], 16, 32)
+			if err != nil {
+				return nil
+			}
+			u.BCD = int(val)
+		case "DEVNAME":
+			u.DevicePath = "/dev/" + values[1]
+		default:
+			log.Log.V(2).Infof("Skipping unhandled line: %s", line)
+		}
+	}
+	return &u
+}
+
+type LocalDevices struct {
+	devices []*USBDevice
+}
+
+// finds by vendor and product
+func (l *LocalDevices) find(vendor, product int) *USBDevice {
+	for _, local := range l.devices {
+		if local.Vendor == vendor && local.Product == product {
+			return local
+		}
+	}
+	return nil
+}
+
+// remove all cached elements
+func (l *LocalDevices) remove(cache map[string]*USBDevice) {
+	devices := []*USBDevice{}
+	for _, local := range l.devices {
+		if _, exists := cache[local.GetID()]; !exists {
+			devices = append(devices, local)
+		}
+	}
+	l.devices = devices
+}
+
+func discoverPluggedUSBDevices() *LocalDevices {
+	usbDevices := make([]*USBDevice, 0)
+	err := filepath.Walk(PathToUSBDevices, func(path string, info os.FileInfo, err error) error {
+		// Ignore named usb controllers
+		if strings.HasPrefix(info.Name(), "usb") {
+			return nil
+		}
+		// We are interested in actual USB devices information that
+		// contains idVendor and idProduct. We can skip all others.
+		if _, err := os.Stat(filepath.Join(path, "idVendor")); err != nil {
+			return nil
+		}
+
+		// Get device information
+		if device := parseSysUeventFile(path); device != nil {
+			usbDevices = append(usbDevices, device)
+		}
+		return nil
+	})
+
+	if err != nil {
+		log.Log.Reason(err).Error("Failed when walking usb devices tree")
+	}
+	return &LocalDevices{devices: usbDevices}
+}
+
+func parseSelector(s *v1.USBSelector) (int, int, error) {
+	val, err := strconv.ParseInt(s.Vendor, 16, 32)
+	if err != nil {
+		return -1, -1, err
+	}
+	vendor := int(val)
+
+	val, err = strconv.ParseInt(s.Product, 16, 32)
+	if err != nil {
+		return -1, -1, err
+	}
+	product := int(val)
+
+	return vendor, product, nil
+}
+
+func discoverAllowedUSBDevices(usbs []v1.USBHostDevice) map[string][]*PluginDevices {
+	// The return value: USB Device Plugins found and permitted to be exposed
+	plugins := make(map[string][]*PluginDevices)
+	// All USB devices found plugged in the Node
+	localDevices := discoverPluggedUSBDevices()
+	for _, usbConfig := range usbs {
+		resourceName := usbConfig.ResourceName
+		count := 0
+		// Tries the same resource name multiple times.
+		for true {
```

Comment:
To keep matching the resource-name selector with available devices. I'm not sure if your question is about the code (e.g: the logic could be better) or the rationale behind the comment + git commit log. Let me quote the interesting part from the later here:

```
     E1: - resourceName: kubevirt.io/storage
           selectors:
             - vendor: "46f4"
               product: "0001"

         Defines a single USB device per k8s Device. If node01 has 3 such
         USB devices, "kubevirt.io/storage" will expose 3 k8s Devices for
         node01
```

So, the `for` is to iterate over resource-name while it can. I wish Go had a `while`, but `for` it is :)

Reply:
For the record, I'm reworking this part and moving some of the logic inside `localDevices`, the result should be something like this..

```Go
func discoverAllowedUSBDevices(usbs []v1.USBHostDevice) map[string][]*PluginDevices {
	// The return value: USB Device Plugins found and permitted to be exposed
	plugins := make(map[string][]*PluginDevices)
	// All USB devices found plugged in the Node
	localDevices := discoverLocalUSBDevicesFunc()
	for _, usbConfig := range usbs {
		resourceName := usbConfig.ResourceName
		index := 0
		for usbdevs, foundAll := localDevices.fetch(usbConfig.Selectors); foundAll; {
			// Create new USB Device Plugin with found USB Devices for this resource name
			pluginDevices := newPluginDevices(resourceName, index, usbdevs)
			plugins[resourceName] = append(plugins[resourceName], pluginDevices)
			index++
		}
	}
	return plugins
}
```